content/blog/2022-05-10-compcert/index.md

```diff
@@ -0,0 +1,112 @@
+
+The author provides five ways to think about a compiler from a semantic perspective:
+
+1. $ \forall B, S \Downarrow B \Leftrightarrow C \Downarrow B $
+2. $ S \text{ safe} \Rightarrow (\forall B, C \Downarrow B \Rightarrow S \Downarrow B) $
+3. $ \forall B \notin \text{Wrong}, S \Downarrow B \Rightarrow C \Downarrow B $ 
+4. $ S \vDash \text{Spec} \Rightarrow C \vDash \text{Spec} $
+5. $ S \text{ safe} \Rightarrow C \text{ safe} $ 
+
+
+These are not interchangeable, but the author explains how they relate to each other and points out features (such as transitivity) that they have in common. In class we took a deeper look at $ S \vDash \text{Spec} \Rightarrow C \vDash \text{Spec} $. While CompCert does not explicitly use that flavor of semantic preservation in its proof, this is arguably the easiest way to appreciate CompCert’s overall goal.
+
+The author surveys three ways to arrive at this trust.
+
+1. Verify the compiler:
+<p align="center">
+    $ \forall S, C, Comp(S) = \text{OK}(C) \Rightarrow S \approx C $
+</p>
+
+2. Don’t trust the compiler. Verify the validator:
+<p align="center">
+    $ \forall S, C, Validate(S, C) = \text{true} \Rightarrow S \approx C $
+</p>
+
+3. Don’t trust the (certificate-generating) compiler. Verify the certificate-checker:
+<p align="center">
+    $ \forall S, C, Check(S, C, Spec, \pi) = \text{true} \Rightarrow S \approx C $
+</p>
+
+Again, these are not interchangeable. They share a deep connection and are, from a formal perspective, equally convincing. This was confusing to the class, with a bit of a sentiment that a certificate-generating untrusted compiler combined with a verified certificate-checker was the best option. We spent some time clarifying this confusion.
+
+For the most part CompCert makes one choice of “trust” and achieves this trust in one way:
+<p align="center">
+    $ \forall S, C, B \notin \text{Wrong}, Comp(S) = \text{OK}(C) \land S \Downarrow B \Rightarrow C \Downarrow B $
+</p>
+
+However, it uses the untrusted code + verified validator approach in one step, and we discussed the merits of this design decision in class. 
+
+
+### Overview of the CompCert Compiler
+
+<p align="center">
+<img src="fig1.png" alt="alt_text" title="image_tooltip" style="zoom:25%;" />
+</p>
+
+The CompCert compiler is composed of 14 passes that go through 8 intermediate languages.
+The source language of CompCert is Clight, a large subset of the C programming language. Although Clight omits several C features such as extended-precision arithmetic and goto statements, it turns out to be sufficient for writing critical embedded software. The target language of CompCert is PPC, a subset of the PowerPC assembly language. This translation from Clight to PPC is formally verified. However, some parts of the compiler are not verified yet including parsing, assembling, and linking. CompCert's intermediate languages are described below in more detail.
```

Comment:
Maybe put stuff like the PPC target in past tense, since you acknowledge elsewhere that this _was_ true at the point this CACM article came out but has since changed (a lot), and CompCert is still a going concern (unlike many research systems).